### DISCUSSION TITLE
Try

### DISCUSSION THREAD

**Discussion by lacasaprivata2** [bitwizeshift/result#4] [Ideas]
Add a TRY macro that returns error otherwise continues

**Comment by bitwizeshift** [comment]
Hi @vitruvvius, thanks for making this suggestion. My apologies for taking so long to reply to this; I've had this project on the backburner for a little while.

As a fan of Rust, I do see the value in what you are suggesting -- but in general I'm against introducing any macro-based solution into this library. With a term as generic as `TRY`, this is also quite a delicate situation that could cause issues with adoption. I've seen several legacy libraries that have `TRY` macros already defined for different purposes, and redefining it in this library could prevent potential adoption due to conflicts.

One option would be to introduce something more verbose like `RESULT_TRY` -- but I'm not a huge fan of this naming or this approach. 
On the surface, I don't think implementing a `TRY` in user code should be all that complicated, and so my general thought has been to leave this as an opt-in feature for the consumer. That said, I could be convinced to introduce this as an "extra" in a separate header for it to be opt-in if there's enough demand.

----

If `TRY` were to be added, one thing that needs to be established is what exactly the behavior is. In particular, does `TRY(x)` return `x.error()` (which returns just the error), or does it return `cpp::fail(x.error())` (which converts to a new `result`)? Part of my rationale for leaving this out is that this is easier for the consumer to decide than it is to provide a generic solution for this. 

I'm open to any suggestions!
